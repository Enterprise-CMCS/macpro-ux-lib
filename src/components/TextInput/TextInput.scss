@use "../../sass/macbis-colors" as c;

.required-star {
  color: c.$error;
}

.usa-input-group input:focus {
  outline: 0.25rem solid c.$focus;
}
