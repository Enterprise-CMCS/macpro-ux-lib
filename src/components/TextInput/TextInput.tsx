import React, { forwardRef } from "react";

type InputElements = JSX.IntrinsicElements["input"];

interface Props extends InputElements {
  id: string;
  label: string;
  errorMessage?: string;
  inputError?: boolean;
  inputSuccess?: boolean;
  prefix?: string;
  required?: boolean;
  suffix?: string;
}

/**
 * TextInput Component
 * @param {string}  id             A unique identifier for the input.
 * @param {string}  [errorMessage] Error message text displayed when inputError === true.
 * @param {boolean} [inputError]   Triggers error message and error styling.
 * @param {boolean} [inputSuccess] Trigger success styling.
 * @param {string}  [prefix]       Text to be displayed at the front of input field. Not stored in value. Ex: currency indicator.
 * @param {boolean} [required]     Adds semantic required attr and appends an * to the end of the input label.
 * @param {string}  [suffix]       Text to be displayed at the end of input field. Not stored in value. Ex: mass indicator (lbs, fl oz)
 */

export const TextInput = forwardRef<HTMLInputElement, Props>(function TextInput(
  props,
  ref
) {
  const {
    id,
    label,
    errorMessage,
    inputError = false,
    inputSuccess = false,
    prefix,
    required = false,
    suffix,
    ...otherProps
  } = props;

  return (
    <div
      className={
        "usa-form-group " +
        (errorMessage || inputError ? "usa-form-group--error" : "")
      }
    >
      <label
        className={
          "usa-label " + (errorMessage || inputError ? "usa-label--error" : "")
        }
        htmlFor={id}
      >
        {label}
        {required && <span className="required-star">*</span>}
      </label>

      {errorMessage && (
        <span className="usa-error-message" role="alert">
          {errorMessage}
        </span>
      )}

      <div
        className={
          "usa-input-group" +
          (errorMessage || inputError ? " usa-input-group--error" : "") +
          (inputSuccess ? " usa-input--success" : "")
        }
      >
        {prefix && (
          <span className="usa-input-prefix" aria-hidden="true">
            {prefix}
          </span>
        )}
        <input
          aria-describedby={`${inputError ? "input-error-message" : ""}`}
          className={`usa-input`}
          id={`input-type-text-${id}`}
          name={fieldName}
          onBlur={() => setFocused(false)}
          onChange={handleChange}
          onFocus={() => setFocused(true)}
          required={required}
          value={inputValue}
          {...rest}
        />
        {suffix && (
          <span className="usa-input-suffix" aria-hidden="true">
            {suffix}
          </span>
        )}
      </div>
    </div>
  );
};
