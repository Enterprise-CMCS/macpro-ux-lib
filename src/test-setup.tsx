import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
