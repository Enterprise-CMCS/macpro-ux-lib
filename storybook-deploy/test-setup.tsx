import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };

/**
 * Clean up attributes from the rendered component by removing the generated ids.
 *
 * Before:
 *   for="input-type-text-123456"
 *   id="input-type-text-123456"
 *
 * After:
 *   for="input-type-text"
 *   id="input-type-text"
 *
 * This is useful during snapshot testing. Snapshot comparisons will fail without addressing these items.
 *
 * @param {HTMLElement} container The component to clean up.
 * @param {prefixes}      string[]  The reference substrings to leave alone.
 */
