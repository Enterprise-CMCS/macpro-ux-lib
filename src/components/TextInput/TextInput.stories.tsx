import React from "react";
import { StoryFn, Meta } from "@storybook/react";

import { TextInput } from "./TextInput";

export default {
  title: "COMPONENTS/TextInput",
  component: TextInput,
  args: {
    label: "Input Label",
    fieldName: "input-type-text",
    inputError: false,
    inputSuccess: false,
    required: false,
  },
  argTypes: {
    label: {
      description: "Field label.",
    },
    fieldName: {
      description: "Name of the input field.",
    },
    errorMessage: {
      description: "Error message text displayed when inputError === true.",
    },
    inputError: {
      description: "Triggers error message and error styling.",
    },
    inputFilter: {
      description:
        "Used to limit input values. If a RegExp is not provided, all input types are allowed.",
    },
    inputSuccess: {
      description: "Trigger success styling.",
    },
    prefix: {
      description:
        "Text to be displayed at the front of input field. Not stored in value. Ex: currency indicator.",
    },
    required: {
      description:
        "Adds semantic required attr and appends an * to the end of the input label.",
    },
    suffix: {
      description:
        "Text to be displayed at the end of input field. Not stored in value. Ex: mass indicator (lbs, fl oz)",
    },
  },
} as Meta<typeof TextInput>;

const Template: StoryFn<typeof TextInput> = ({ ...rest }) => (
  <TextInput {...rest} />
);

export const Default = Template.bind({});
export const PrefixSuffixAndFilter = Template.bind({});
export const RequiredAndError = Template.bind({});
export const Success = Template.bind({});
Default.args = {};

PrefixSuffixAndFilter.args = {
  inputFilter: /^-?\d*$/i,
  label: "This Field Only Accepts Numbers",
  prefix: "#",
  suffix: "lbs.",
};

RequiredAndError.args = {
  errorMessage: "Helpful Error Message",
  inputError: true,
  label: "Required Input Field",
  required: true,
  id: "required-and-error-textarea",
};

Success.args = {
  label: "Field with Success Indicator",
  inputSuccess: true,
  id: "success-textarea",
};
