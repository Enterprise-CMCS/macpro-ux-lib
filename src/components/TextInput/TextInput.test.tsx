import React from "react";
import { screen, render, getByLabelText, getByText } from "../../test-setup";
import fireEvent from "@testing-library/user-event";
import { TextInput } from "./TextInput";

describe("TextInput", () => {
  describe("default", () => {
    const errorMessage = "This is an example of an error message.";

    beforeEach(() => {
      render(
        <TextInput
          id="test-input"
          label="Testing Input"
          name="testing-input"
          errorMessage={errorMessage}
        />
      );
    });

    it("renders", () => {
      const comp = screen.getByLabelText("Testing Input");
      expect(comp).toBeInTheDocument();
    });

    it("has no initial value", () => {
      const comp = screen.getByLabelText("Testing Input");
      expect(comp).toHaveDisplayValue("");
    });

    it("accepts any value", () => {
      const comp = screen.getByLabelText("Testing Input");
      fireEvent.type(comp, "abc123!@#");
      expect(comp).toHaveDisplayValue("abc123!@#");
    });

    it("does not show the errorMessage", () => {
      const comp = screen.getByLabelText("Testing Input");
      expect(comp).toBeInTheDocument();
      expect(screen.queryAllByText(errorMessage).length).toBe(0);
    });
  });

  it("should show an error message", () => {
    const errorMessage = "This is an example of an error message.";
    render(
      <TextInput
        id="test-input"
        label="Testing Input"
        name="testing-input"
        errorMessage={errorMessage}
        inputError
      />
    );
    expect(screen.getByText(errorMessage)).toBeInTheDocument();
  });

  it("should be required", () => {
    render(
      <TextInput
        id="test-input"
        label="Testing Input"
        name="testing-input"
        required
      />
    );
    const comp = screen.getByLabelText("Testing Input*");
    expect(comp).toBeInTheDocument();
    comp.hasAttribute("required");
  });

  describe("snapshots", () => {
    it("renders the default TextInput", () => {
      const { container } = render(
        <TextInput id="test-input" label="Testing Input" name="testing-input" />
      );
      expect(container).toMatchSnapshot();
    });

    it("renders a required TextInput", () => {
      const { container } = render(
        <TextInput
          id="test-input"
          label="Testing Input"
          name="testing-input"
          required
        />
      );
      expect(container).toMatchSnapshot();
    });

    it("renders an error", () => {
      const errorMessage = "This is an example of an error message.";
      const { container } = render(
        <TextInput
          id="test-input"
          label="Testing Input"
          name="testing-input"
          errorMessage={errorMessage}
        />
      );
      expect(container).toMatchSnapshot();
    });

    it("renders the success state", () => {
      const { container } = render(
        <TextInput
          id="test-input"
          label="Testing Input"
          name="testing-input"
          inputSuccess
        />
      );
      expect(container).toMatchSnapshot();
    });

    it("renders the prefix and suffix", () => {
      const { container } = render(
        <TextInput
          id="test-input"
          label="Testing Input"
          name="testing-input"
          prefix="$"
          suffix="lbs."
        />
      );
      expect(container).toMatchSnapshot();
    });
  });
});
