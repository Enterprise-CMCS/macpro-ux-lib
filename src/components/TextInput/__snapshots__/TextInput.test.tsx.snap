// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput snapshots renders a required TextInput 1`] = `
<div>
  <div
    class="usa-form-group "
  >
    <label
      class="usa-label "
      for="test-input"
    >
      Testing Input
      <span
        class="required-star"
      >
        *
      </span>
    </label>
    <div
      class="usa-input-group"
    >
      <input
        class="usa-input"
        id="test-input"
        name="testing-input"
      />
    </div>
  </div>
</div>
`;

exports[`TextInput snapshots renders an error 1`] = `
<div>
  <div
    class="usa-form-group "
  >
    <label
      class="usa-label "
      for="test-input"
    >
      Testing Input
    </label>
    <div
      class="usa-input-group"
    >
      <input
        class="usa-input"
        id="test-input"
        name="testing-input"
      />
    </div>
  </div>
</div>
`;

exports[`TextInput snapshots renders the default TextInput 1`] = `
<div>
  <div
    class="usa-form-group "
  >
    <label
      class="usa-label "
      for="test-input"
    >
      Testing Input
    </label>
    <div
      class="usa-input-group"
    >
      <input
        class="usa-input"
        id="test-input"
        name="testing-input"
      />
    </div>
  </div>
</div>
`;

exports[`TextInput snapshots renders the prefix and suffix 1`] = `
<div>
  <div
    class="usa-form-group "
  >
    <label
      class="usa-label "
      for="test-input"
    >
      Testing Input
    </label>
    <div
      class="usa-input-group"
    >
      <span
        aria-hidden="true"
        class="usa-input-prefix"
      >
        $
      </span>
      <input
        class="usa-input"
        id="test-input"
        name="testing-input"
      />
      <span
        aria-hidden="true"
        class="usa-input-suffix"
      >
        lbs.
      </span>
    </div>
  </div>
</div>
`;

exports[`TextInput snapshots renders the success state 1`] = `
<div>
  <div
    class="usa-form-group "
  >
    <label
      class="usa-label "
      for="test-input"
    >
      Testing Input
    </label>
    <div
      class="usa-input-group usa-input--success"
    >
      <input
        class="usa-input"
        id="test-input"
        name="testing-input"
      />
    </div>
  </div>
</div>
`;
